### DISCUSSION TITLE
Problem with IP addresses in Clusters with private and public network

### DISCUSSION THREAD

**Discussion by simonostendorf** [syself/cluster-api-provider-hetzner#834] [Q&A]
Hello,

I am wondering what I am doing wrong when I create a cluster.

Here is the initial scenario:
I want a cluster where the nodes have both private and public IP addresses. The whole thing is currently created via a ClusterClass. 

Problem:
The nodes get a private and public address as desired, but the pods partly use the public IP. Can I pin this to the private IP? I can't enter this in the configs because I don't know the IP there and ClusterAPI would then have to replace it. 

Thank you for your help!

![image](https://github.com/syself/cluster-api-provider-hetzner/assets/70595012/4f0bc91d-3d95-4794-a5ec-672339d4aaf8)

Edit: I am using advertise-address 10.0.0.2 (API LB) and bind-address 0.0.0.0

The pod (eg. kube-apiserver) shows `hostIP: 10.0.0.9`, `podIP: 128.140.X.X` and `podIPs: [ip: 128.140.X.X]`.

I could fix this first error by adding the node private ip together with the hostname into the `/etc/hosts`. But kubernetes service for api (ip 10.20.0.1 for me) doesn't work because the endpoints use the public ip because kubeadm uses this is the advertise address [refering to this post](https://serverfault.com/a/1029239)

**Comment by batistein** [comment]
This needs to be configured in the cni. The cluster-api is not responsible for this. 
FYI: the cluster-api only manages the cluster lifecycle, configuration of the cluster needs to be done in the node-image, kubeadm or in the basic applications / cluster-addons (e.g cni, ccm etc.)
